### PR TITLE
Add variants to petsc

### PR DESCRIPTION
**Pull request by glennpj**
This PR adds the follolwing variants to the petsc package

- gmp
- jpeg
- libpng
- giflib
- mpfr
- netcdf
- pnetcdf (parallel-netcdf)
- moab
- eigen
- random123
- exodusii
- mstk
- cgns
- memkind
- muparser
- p4est
- saws
- libyaml
- zstd

### CODE REVIEW

**Comment by adamjstewart** [comment]
@balay @barrysmith @jedbrown can you review this?

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: We'll have to trim this list as some of these packages are primarily dependencies of dependencies.

For ex: zlib is a dependency of hdf5 - so its not listed as a variant - but as:
```
    depends_on('zlib', when='+hdf5')
```
In this case gmp is a dependency of mpfr  [which hardly gets tested].

BTW: Are these changes tested - i.e do spack builds work with these variants enabled?

Comment (glennpj) [reply var/spack/repos/builtin/packages/petsc/package.py]: It is possible to have petsc build with gmp and not with mpfr, however, you are correct, building with mpfr requires gmp to be enabled. That is not reflected in the initial PR so would need to be corrected. I did not add a zlib variant for two reasons,
1. It is already the "default" because `+hdf5` is set by default.
1. It may be that zlib should be enabled by default and set as a dependency, but I had not explored that.

With regard to testing, our site's non-spack built petsc-3.10.3 was built including the following variants in this PR:
- netcdf
- mpfr
- libpng
- libjpeg
- gmp

My initial motivation for this PR was to add netcdf support. I then added the other things that I built previous versions of petsc with, and finally added others while I was in there. If any of those seem unneeded/unuseful/silly then I can pull them out. My main desire is for netcdf support, and to a lesser degree, parallel-netcdf.

Comment (jedbrown) [reply var/spack/repos/builtin/packages/petsc/package.py]: I think you've missed @balay's point. PETSc does not use GMP. PETSc's configure knows about it so it can find/build it as a dependency for MPFR (which is used in one rarely-used function).

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Note: the premise here is: we need variants only for packages that PETSc directly interfaces with.

The reference to hf5, zlib is an example of the case of a dependency of a dependency [as currently implemented in spack]

Also note: PETSc does not have an interface to netcdf  parallel-netcdf. They are dependencies for other pacakges. (i.e a varaint should not be added for them)

* pnetcdf is a dependency for exodusii, netcdf (optional)
* netcdf is a dependency for ADIOS (optional),moab (optional),Trilinos (alternate way of installing exodus,ML etc)

And generally - with shared libraires - one doesn't need to worry about dependencies of dependencies. But in static builds we do - hence this complexity in petsc configure. [apart from the fact that petsc configure can build these dependencies on its own]

Comment (glennpj) [reply var/spack/repos/builtin/packages/petsc/package.py]: Well, petsc can build with and link gmp independently of mpfr. Whether it uses it is another story, but,
```
--with-gmp=1 --with-mpfr=0
```
will produce a petsc library linked with libgmp. In any event, if people think it is not a necessary variant, it can be removed.

Comment (jedbrown) [reply var/spack/repos/builtin/packages/petsc/package.py]: I would call this a bug in PETSc configure that isn't worth "fixing", except perhaps to convince people who want to turn "everything" on. It's an artifact of PETSc's configure being a sort of package manager. It's easy to check that Satish and I aren't lying to you about PETSc not using GMP. If you add this as a Spack variant, it would only serve to mislead more people.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: To illustrate:
```
balay@sb /home/balay/junk
$ cat sz.c
#include <stdio.h>
int main()
{
printf("test\n");
return 0;
}
balay@sb /home/balay/junk
$ gcc sz.c
balay@sb /home/balay/junk
$ ldd a.out
	linux-vdso.so.1 (0x00007ffcae5ef000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f683e203000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f683e3f5000)
balay@sb /home/balay/junk
$ mpicc sz.c
balay@sb /home/balay/junk
$ ldd a.out
	linux-vdso.so.1 (0x00007ffcfd7e8000)
	libmpi.so.12 => /home/balay/soft/mpich-3.3.2/lib/libmpi.so.12 (0x00007f9e199af000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f9e197bf000)
	libgfortran.so.5 => /lib64/libgfortran.so.5 (0x00007f9e194f7000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f9e193b1000)
	libOpenCL.so.1 => /lib64/libOpenCL.so.1 (0x00007f9e1938f000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f9e1936d000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f9e19360000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f9e19345000)
	libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007f9e192fb000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9e19ef2000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f9e192f4000)
```

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: i.e just because `libmpi.so.12` is linked in does not imply the above test code is dependent on MPI

Comment (glennpj) [reply var/spack/repos/builtin/packages/petsc/package.py]: I believe you, and always did, I was just presenting what I was observing. It sounds to me like gmp and mpfr variants can be nixed.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]:
```
$ git grep PETSC_HAVE_MPFR
src/dm/dt/interface/dt.c:#if defined(PETSC_HAVE_MPFR)
src/dm/dt/interface/dt.c:#if defined(PETSC_HAVE_MPFR)
src/dm/dt/tests/ex3.c:#if defined(PETSC_HAVE_MPFR)
```
gmp should be nixed. mpfr can be kep [however configure will insist on passing --with-gmp option for -with-mpfr to work]

netcdf, pnetcdf variants should be nixed aswell.

Will check the complete list...

Comment (glennpj) [reply var/spack/repos/builtin/packages/petsc/package.py]: I just saw the most recent comment from @balay. I am somewhat confused by the following:
>Note: the premise here is: we need variants only for packages that PETSc directly interfaces with.
...
Also note: PETSc does have an interface to netcdf parallel-netcdf. They are dependencies for other pacakges. (i.e a varaint should not be added for them)

The first statement indicates that a variant would be viable if there is an interface but the second statement seems to contradict that. Or, did you mean that, "PETSc does **not** have an interface to netcdf parallel-netcdf"?

Sorry, I seem to be learning that I know even less about PETSc than I thought.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Sorry - typo. `petsc does not have an interface to netcdf, parallel-netcdf`

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: * gmp     - nixed
* jpeg
* libpng   (configure checks for zlib as dependencies)
* giflib
* mpfr	   (configure checks for gmp as dependencies)
* netcdf  - nixed
* pnetcdf - nixed
* moab
* eigen   - nixed
* random123
* exodusii (configure checks for hdf5, netcdf, pnetcdf as dependencies)
* mstk    - nixed
* cgns     (configure checks for hdf5 as dependencies)
* memkind
* muparser - nixed (don't know why we have this)
* p4est    (configure checks for zlib as dependencies)
* saws
* libyaml
* zstd	   - nixed (don't know why we have this)

Comment (glennpj) [reply var/spack/repos/builtin/packages/petsc/package.py]: > i.e just because `libmpi.so.12` is linked in does not imply the above test code is dependent on MPI

Yes, I understand that. I was trusting the configuration mechanism to use what it was looking for, which was apparently, a mistake.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Basically configure is not rejecting mpicc when user specifies it [even-though its not required in this specific case]. And it trusts the user need it for other reasons.. [and checks if things work with it or not]

Comment (jedbrown) [reply var/spack/repos/builtin/packages/petsc/package.py]: `zstd` can be dropped here. I've been using it in a branch with a student, but that feature hasn't landed yet.

Comment (glennpj) [reply var/spack/repos/builtin/packages/petsc/package.py]: So, the missing piece for me was not understanding there was a distinction between packages presented in `configure --help` that interfaced with PETSc and their dependencies. It would be nice if that could be made more clear in the `configure --help` output so that people who are not intimately familiar with PETSc, but just trying to package it for their users, would have more guidance.

The question that I have at this point is should I pursue adding the options from the list that @balay curated to the spack package or just drop this PR?

Comment (jedbrown) [reply var/spack/repos/builtin/packages/petsc/package.py]: The information is in the configure files; see `self.useddirectly = 0` for those that are only indirect dependencies.
```console
$ git grep useddirecly
```
Yes, someone could make that visible in the `./configure --help` output. Preferably not more lines; the output is long enough already.

**Comment by glennpj** [comment]
Restart tests.

**Comment by glennpj** [comment]
@adamjstewart When you get a chance could you take a look at this PR to see if I addressed everything.

**Review by adamjstewart — APPROVED** [review]
Looks fine to me, but I don't know anything about `petsc`. Waiting for one of the maintainers to approve before merging.

**Comment by glennpj** [comment]
@jedbrown and/or @balay Could you take a look at this to see if everything has been addressed?

Thanks.

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py:147]: This is tricky.

petsc can build with hdf5\~mpi, mumps\~mpi.

And the other packages (exodusii cgns) have dependencies on hdf5 [its not clear if we tested them with hdf5\~mpi]. So I'm not sure what's appropriate here.

I see this issue is with current code in spack - and not related to this MR. [its just moving the check]. So I guess its fine to keep these conflicts for now - and update once we can test those combinations in spack

**Review by balay — APPROVED** [review]

**Comment by balay** [comment]
@glennpj thanks for the improvements! sorry about the delay in response

**Comment by glennpj** [comment]
> @glennpj thanks for the improvements! sorry about the delay in response

No worries. I learned quite a bit working on this PR and that is always a good thing.

**Pull request merged**
